go to prod check list: reworked chapters order and add some new content (#5292)
solo-io/gloo: docs/content/operations/production_deployment/_index.md

This document shows some of the Production options that may be useful. We will continue to add to this document and welcome users of Gloo Edge to send PRs to this as well.

> ## Securing the control-plane from configuration errors

Reviewer: ```suggestion
## Safeguarding the control plane configuration
```

Example:

```yaml
gloo:
  settings:
    invalidConfigPolicy:
      invalidRouteResponseBody: Gloo Gateway has invalid configuration. Administrators
        should run `glooctl check` to find and fix config errors.
      invalidRouteResponseCode: 404
      replaceInvalidRoutes: true
```

### OpenAPI schema validation on CRDs
Since Gloo v1.8, CRDs are given OpenAPI schemas. This way, the kube `api-server` (validation webhook) will refuse _Custom Resources_ with invalid definitions.

### More safeguards
In addition to the CRD schemas, Gloo can perform a deeper inspection of the _Custom Resources_.

You can use these flags:

```yaml
gloo:
  gateway:
    validation:
      allowWarnings: false # reject if warning status or rejected status
      alwaysAcceptResources: false # reject invalid resources
      warnRouteShortCircuiting: true
```

## Performance tips

### Disable Kubernetes destinations

> Gloo Edge out of the box routes to upstreams. It can also route directly to Kubernetes destinations (bypassing upstreams). Upstreams is the recommended abstraction to which to route in VirtualServices, and you can disable the Kubernetes destinations with the `settings.disableKubernetesDestinations`. This saves on memory overhead so Gloo Edge pod doesn't cache both upstreams and Kubernetes destinations. 

Reviewer: ```suggestion
Gloo Edge routes to upstreams by default, but it can alternatively be configured to bypass upstreams and route directly to Kubernetes destinations. Because routing to upstreams is the recommended configuration, you can disable the option to route to the Kubernetes destinations with the `settings.disableKubernetesDestinations: true` setting. This setting saves memory because the Gloo Edge pod doesn't cache both upstreams and Kubernetes destinations.
```

You can set this value in the default `Settings` CR by adding the following content:   
```yaml
apiVersion: gloo.solo.io/v1
kind: Settings
metadata:
  name: default
  namespace: gloo-system
spec:
  gloo:
    disableKubernetesDestinations: true
    ...
```

> You can set this value helm overrides by setting:

Reviewer: ```suggestion
You can set this value in your Helm overrides file by adding the following setting:
```

These values can be configured via helm values for the various deployments, such as 
* `gloo.deployment.resources.requests.*` 
* `gatewayProxies.gatewayProxy.podTemplate.resources.requests.*`

See the [helm chart value reference]({{%versioned_link_path fromRoot="/reference/helm_chart_values/" %}}) for a full list.

### Transformations

* Disabling transformation validation can drastically improve the CPU usage on the control-plane side.

```yaml
gloo:
  gateway:
    validation:
      disableTransformationValidation: true # better performances but more risky
```

### Discovery
If you have hundreds of Kubernetes services, you can get better performances on the control-plane if you disable the Upstream and Function discovery.

```yaml
gloo:
  discovery:
    enabled: false
```
You can also patch the `default` *Settings* CR with this value and delete the `discovery` deployment.

### Enable Envoy's gzip filter
Optionally, you may choose to enable Envoy's gzip filter through Gloo Edge. More information on that can be found [here]({{% versioned_link_path fromRoot="/installation/advanced_configuration/gzip/" %}}).

### Set up an EDS warming timeout

> Set up the endpoints warming timeout to some nonzero value. More details [here]({{%versioned_link_path fromRoot="/operations/upgrading/1.3.0/#recommended-settings" %}}).

Reviewer: ```suggestion
Set up the endpoints warming timeout to a non-zero value. More details [here]({{%versioned_link_path fromRoot="/operations/upgrading/1.3.0/#recommended-settings" %}}).
```

## Access Logging

Envoy provides a powerful access logging mechanism which enables users and operators to understand the various traffic flowing through the proxy.
Before deploying Gloo Edge in production, consider enabling access logging to help with monitoring traffic as well as to provide helpful information for troubleshooting.
The [access logging documentation]({{%versioned_link_path fromRoot="/guides/security/access_logging/" %}}) should be consulted for more details.

{{% notice note %}}
Make sure you have the `%RESPONSE_FLAGS%` item in the log pattern. 
{{% /notice %}}

> ## Horizontal scaling - the data-plane

Reviewer: ```suggestion
## Horizontally scaling the data plane
```

### Gateway proxy

> It's fine to scale up the gateway-proxies (Envoy instances). You can use a Deployment or a DeamonSet. The more CPU you will dedicate to the gateway-proxy, the more requests if will be able to handle.

Reviewer: ```suggestion
You can scale up the `gateway-proxies` Envoy instances by using a Deployment or a DeamonSet. The amount of requests that the `gateway-proxies` can process increases with the amount of CPU that the `gateway-proxies` have access to.
```

### Ext-Auth

> You can also scale-up the ExtAuth service. Most of the time, one instance is fine. We have seen deployments with 2 replicas working well.

Reviewer: ```suggestion
You can also scale up the ExtAuth service. Typically, one to two instances are sufficient.
```

> ## Horizontal scaling - the control-plane

Reviewer: ```suggestion
## Horizontally scaling the control plane
```

> ### XDS Relay

Reviewer: same comments as in https://github.com/solo-io/gloo/pull/5212/files

Author: yep, will copy them once the PR 5212 is merged.

{{% /notice %}}

You should consider protecting against control plane downtime (either by node failure or gloo bug) by installing gloo edge alongside the xds-relay helm chart. This helm chart installs a daemonset (configurable, daemonset by default) of xds-relay pods that serve as intermediaries between envoy and gloo (the xds server).

This serves two purposes. One, it separates the lifecycle of gloo edge from the xds cache proxies. This means a failure during helm upgrade will not mean loss of last good xds state. Second, it allows you to scale xds-relay to as many replicas as desired, since gloo is only intended for one replica today. Without xds-relay, if the single gloo replica were down, then any new envoy proxies spun up would be unable to get valid configuration.

To enable:

- install `xds-relay`:
  - `helm repo add xds-relay https://storage.googleapis.com/xds-relay-helm`
  - `helm install xdsrelay xds-relay/xds-relay`
- install gloo edge with the following helm values for each proxy (envoy) to point them towards xds-relay:
```yaml
gatewayProxies:
  gatewayProxy:
    xdsServiceAddress: xds-relay.default.svc.cluster.local
    xdsServicePort: 9991
```

### Other components

> For the other control plane components, such as the Gateway deployment or the Discovery deployment, *DO NOT* scale them. Today, it can lead to race conditions and there are no need to scale them up.

Reviewer: ```suggestion
*DO NOT* scale other control plane components, such as the Gateway deployment or the Discovery deployment. Scaling these components provides no benefit and can lead to race conditions.
```

### Upstream health checks

In addition to defining health checks for Envoy, you should strongly consider defining health checks for your `Upstreams`.
These health checks are used by Envoy to determine the health of the various upstream hosts in an upstream cluster, for example checking the health of the various pods that make up a Kubernetes `Service`. This is known as "active health checking" and can be configured on the `Upstream` resource directly.
[See the documentation]({{% versioned_link_path fromRoot="/guides/traffic_management/request_processing/upstream_health_checks/" %}}) for additional info.

### Other considerations

Additionally, "outlier detection" can be configured which allows Envoy to passively check the health of upstream hosts.
A helpful [overview of this feature is available in Envoy's documentation](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/outlier).
This can be configured via the `outlierDetection` field on the `Upstream` resource. See the {{< protobuf name="gloo.solo.io.Upstream" display="API reference for more detail" >}}.

> Also, consider using `retries` on your _routes_. The default value for this attribute is `1`. Generally spealing, we observed better results with the value being set to `3`.

Reviewer: ```suggestion
Also, consider using `retries` on your _routes_. The default value for this attribute is `1`, which you can increase to `3` for better results.
```

## Metrics and monitoring

### Grafana dashboards

> You will find a dedicated Grafana dashboard for each `Upstream` _Custom Resource_ that exists. These dashboard are pretty handy when you want to understand if your active/passive health-checks are working, as well as retries, requests rate and latency, responses codes, number of active connections and more.

Reviewer: ```suggestion
A dedicated Grafana dashboard exists for each `Upstream` _Custom Resource_. These dashboards can help you verify whether your active/passive health checks are working, as well as provide insight into retries, requests rate and latency, responses codes, number of active connections, and more.
```
